Payjoin-spec.md

```diff
@@ -0,0 +1,182 @@
+
+### Preventing mempool replacement
+A safe way to implement payjoin, is for both the sender and receiver to try broadcasting the original transaction at some fixed interval period regardless of the state of the payjoin.
+
+If the receiver was not properly adding fees to the payjoin transaction, the original transaction would end up replacing the payjoin transaction in the mempool.
+
+### Receiver does not need to be a full node
+Because the receiver need to bump the keep the same fee rate than the original PSBT, it needs the input's UTXO information to know what is the original fee rate. Without PSBT, light wallets like Wasabi Wallet would not be able to receive a payjoin transaction.
+
+The validation (policy and consensus) of the original transaction is optional, a receiver without a full node can decide to create the payjoin transaction and automatically broadcast the original transaction after a timeout 1 minute, and only verifying that it has been propagating in the network.
+
+However, automated systems (like BTCPay Server) need to verify the transaction to prevent UTXO probing attacks. 
+
+This is not a concern for receivers using end-user wallets such as Wasabi Wallet, such receiver can just limit the number of original PSBT proposal a specific address to one. With such wallet, the attacker has no way to generate new deposit addresses to probe the UTXOs.
+
+## Implementation notes
+
+### Receiver side
+
+The BTCPayServer implementation is making the following checks on the original PSBT:
+
+* The original PSBT must be finalized (Fails with error `psbt-not-finalized`)
+* All the inputs must be P2WPKH (Fails with error `unsupported-inputs`) [*](#unsupported-inputs)
+* The PSBT is sane (Fails with error `insane-psbt`)
+* All inputs have the `witnessUTXO` (Fails with error `need-utxo-information`)
+* Check that there is no global xpubs, HD Key or public key information in the PSBT (Fails with error `leaking-data`)
+* Check that the extracted transaction can be accepted by mempool (Fails with error `invalid-transaction`)
+
+Then the original PSBT is deemed valid, however, other checks are done which depends on the state of the receiver's state.
+
+* Check that it pays an actual invoice (Fails with error `invoice-not-found`)
+* Check that the invoice does not have any outstanding payment (Fails with error `already-paid`)
+* Check that the whole amount of the invoice is paid (Fails with error `invoice-not-fully-paid`)
+* Check that the inputs used in the original PSBT have never been used for another payjoin (Fails with `inputs-already-used`)
+* Check that the receiver has available UTXO to contribute to the payjoin (Fails with error `out-of-utxos`)
+* Check that the receiver can actually sign (Fails with error `unavailable`)
+
+Finally, we need to increase the fees to pay for the increased size of payjoin transaction.
+From this point, the receiver will automatically try to broadcast the original transaction after 1 minute, whatever happens.
```

Comment:
Even if the payjoin validation fails (except for transaction being invalid for mempool or invoice is payable in full), don't we broadcast the tx at the receiver side instantly?

Reply:
we only do for `out-of-utxos`